### ISSUE TITLE
Package MirageOS

### ISSUE THREAD

**Issue by copumpkin** [NixOS/nixpkgs#23955]
I'd love to be able to play with [MirageOS](https://mirage.io/) and we have a decent amount of the underlying OCaml machinery and libraries already packaged. Would be cool to make carefully tailored MirageOS images inside Nix.

cc @maurer @vbgl as OCaml users I know of

**Comment by copumpkin** [comment]
I've been doing this while doing something else. Will push up what I have soon.

**Comment by copumpkin** [comment]
This is done but a bit messy on my local machine. I'll clean it up then push it up later.

**Comment by timbertson** [comment]
I do have mirageOS working under `opam2nix` (https://github.com/timbertson/opam2nix-packages), if that helps. I don't know if/when it'll be ready to merge into nix proper, but it could be a less arduous way of tracking opam upstream.

**Comment by copumpkin** [comment]
Oh that'll be nice. I need it in a non-OCaml package in the short term, unfortunately, so I'll probably merge my work soon regardless, then we can replace it with opam2nix!

**Comment by vbgl** [comment]
Any news on this topic? Thanks.

**Comment by puffnfresh** [comment]
@copumpkin this would help with my work on VPNKit!

**Comment by deliciouslytyped** [comment]
I will be going to Bobkonf this friday (would be nice to meet with any nix-ers! :D ) and there's going to be a tutorial on MirageOS (https://bobkonf.de/2019/mehnert.html). I haven't used any of the ocaml infrastructure yet and I don't know how much work I can put into it over the week, but it would be nice if someone could have another peek at this?

@copumpkin did you manage to post any of your work?

**Comment by deliciouslytyped** [comment]
For the record, I tried just using opam and following https://mirage.io/wiki/install and https://mirage.io/wiki/hello-world ; and that seems to have worked pretty much flawlessly. I just tried running the noop example and there were no problems.

**Comment by armijnhemel** [comment]
I had a very quick chat with Hannes from MirageOS people and he is intrigued. So whoever wants to work on this should contact him, if this is still relevant (because of for example b90908aead3a4bad4d122b3f1d2bfacd0cfd1677 )

**Comment by ehmry** [comment]
I've built the Mirage tool with Nix before: https://github.com/ehmry/mirage-nix

**Comment by Ericson2314** [comment]
Great! I've long wanted to see something along these lines. That would be @hannesm?

Another thing I've long term wanted to see was replacing the few C bits (Solo5, OCaml) with Rust, and coincidentally enough, we're just in the midst of getting better bare metal Rust cross compilation support thanks to @aaronjanse's work for [Redox](https://www.redox-os.org/).

Typically the more obscure toolchains need to be wrangled, the brighter Nix shines, so I figured something like that would be a good opportunity to show of Nix.

**Comment by ehmry** [comment]
Also, Solo5 is in the tree (dd8423f86550ef7a63d704b220d361e977142b29).

**Comment by hannesm** [comment]
Yes, that is me. AFAICT the solo5 toolchain is already packaged in nix, and the mirage utility can be (or is already? -- if not, are there any showstoppers?). Once these pieces are in place, we should think about packaging concrete unikernels into nixOS (I have some ideas about how to achieve that - in the end a configured MirageOS unikernel is an opam package).

So for now, I have two questions:
- is the mirage command line utility (opam package mirage) packaged in nixOS?
- is it easy to package an opam package into nixOS?

**Comment by Ericson2314** [comment]
@hannesm So the slickest way to package opam modules with Nix is to on-demand generate the Nix code using tools like that `opam2nix`. Unfortunately, Nixpkgs doesn't allow generated code on the fly due to an issue with our CI tool (not Nix itself), so we have to vendor the output for opam packages we want to include.

So, to answer your questions

> is it easy to package an opam package into nixOS?

Yes, very.

> is the mirage command line utility (opam package mirage) packaged in nixOS?

It might not be, but it should be no harder than other opam package. Just a matter of articulating the motivation and vendoring the code that's easy to generate.

-------

I think if you/we start with @ehmry's https://github.com/ehmry/mirage-nix, we won't even be blocked on the previous two steps (it also uses `opam2nix`). We can go straight to packaging concrete unikernels into NixOS by writing the requisite NixOS modules in there. Once the whole thing works end-to-end, I think there will be no trouble justifying moving the code to Nixpkgs proper :).

That plan allows iterating with the dynamic Nix code generation and only "baking" it at the end, which should be much more ergonomic.

**Comment by ehmry** [comment]
From what I remember, the dependencies of a Mirage unikernel are determined at configuration time, which means that a NixOS module for something Mirage would break the no-import-from-derivation rule in Nixpkgs. That may require then Mirage stuff to be hosted outside the main tree, but that's less of an inconvenience these days.

Also, `opam2nix` is probably the better way of managing the dependencies than managing them manually. Opam has good metadata so this should be safe to do.

I believe this implies that to realize a configuration written in Nix would require a derivation that calls `mirage` to generate an Opam package that corresponds to the Nix configuration, processing that package with `opam2nix`, and then building the unikernel from that.

**Comment by Ericson2314** [comment]
Yeah, I suspect the Mirage CLI and Nix are going to have a bit of a turf war, but otherwise things should be fine.

**Comment by sternenseemann** [comment]
To see how far we've come and what challenges we still have to overcome, I tried to build the `stub-resolver` unikernel from [roburio/unikernels](https://github.com/roburio/unikernels) today.

I'm pleased to announce that in terms of MirageOS unikernels for the unix target we are basically already there! We are just missing MirageOS libraries here and there, but the fundamentals are there and work fine. My endeavour lead to #107907, #107914 and #107922.

With the following, frankly trivial `shell.nix` (and the PRs merged into a local nixpkgs) it's a matter of `nix-shell --run "mirage configure && mirage build"`:

```shell.nix
let
  pkgs = import <nixpkgs> {};

  op = pkgs.ocamlPackages;

  baseTools = with op; [
    ocaml dune_2 ocamlbuild findlib
  ];

  baseLibraries = with op; [
    mirage mirage-unix mirage-types mirage-types-lwt mirage-logs
    mirage-crypto-rng-mirage mirage-clock-unix mirage-bootvar-unix tcpip
  ];

  unikernelDeps = with op; [
    dns-resolver dns-server logs
  ];
in

pkgs.mkShell rec {
  buildInputs = baseTools ++ baseLibraries ++ unikernelDeps;
}
```

The predicted turf war between `mirage` and the nix infrastructure is also not happening, as it is pretty well behaved and uses standard ocaml tools like `dune`, `ocamlfind` and `ocamlbuild` which we have already figured out.

Here's a screenshot of `stub-resolver` built on NixOS answering a request on my my machine:

![tmp](https://user-images.githubusercontent.com/3154475/103319640-cb073700-4a32-11eb-851d-d12ff4f24fdc.png)

Next steps for MirageOS in nixpkgs as I see it would be:

* [ ] Package missing mirage libraries
* [ ] Add infrastructure for building MirageOS unikernels (should be as lightly altered `buildDunePackage`) and maybe even deploying it
* [ ] Test and figure out other MirageOS targets like Xen

**Comment by Ericson2314** [comment]
@sternenseemann Congrats! Your PRs all looked perfectly fin to me and I merged them.

> The predicted turf war between `mirage` and the nix infrastructure is also not happening, as it is pretty well behaved and uses standard ocaml tools like `dune`, `ocamlfind` and `ocamlbuild` which we have already figured out.

Glad to be wrong so far!

> Test and figure out other MirageOS targets like Xen

...but I suspect once we get into these interesting cross compilation cases, things will emerge :).

**Comment by Ericson2314** [comment]
I wonder if anyone has tried a statically linked Unix target with Musl, since that's a vageuly kindred spirit of a target to Unikernels.

**Comment by sternenseemann** [comment]
> I wonder if anyone has tried a statically linked Unix target with Musl, since that's a vageuly kindred spirit of a target to Unikernels.

Seems like it is planned, but not yet implemented:
<https://github.com/mirage/mirage/issues/880>

Although you can probalby achieve it already by messing around in the build files `mirage configure` generates.

**Comment by sternenseemann** [comment]
MirageOS 4 related upstream issues/PRs:

* https://github.com/mirage/mirage/issues/1217
* https://github.com/Solo5/solo5/pull/494           
* https://github.com/mirage/mirage/pull/1226

**Comment by RyanGibb** [comment]
I've written [hillingar](https://github.com/ryangibb/hillingar) which uses [opam-nix](https://github.com/tweag/opam-nix) to build mirage unikernels. I've written a blog post on the project here: https://ryan.freumh.org/blog/hillingar/

**Comment by asymmetric** [comment]
Good catch, this can be closed.

**Comment by sternenseemann** [comment]
Untrue, MirageOS is not packaged in a meaningful/workable form. Some libraries related to it are packaged, but it is not possible to build/package MirageOS  unikernels.

**Comment by ehmry** [comment]
It's not a matter of missing packages, the problem with Mirage is getting the OPAM configuration stuff working with Nix.

**Comment by RyanGibb** [comment]
> It's not a matter of missing packages, the problem with Mirage is getting the OPAM configuration stuff working with Nix.

Could you expand on this?

**Comment by ehmry** [comment]
The configuration stuff as seen here: https://mirage.io/docs/hello-world

**Comment by RyanGibb** [comment]
> The configuration stuff as seen here: https://mirage.io/docs/hello-world

I'm afraid I don't follow... Could you be more specific about what configuration you forsee having issues working with Nix? I've built multiple mirage unikernels with Nix (see https://github.com/RyanGibb/hillingar/blob/main/src/mirage.nix).

> Does anyone have a list of packages that we'd need to package in order to have meaningful and workable MirageOS packaging? If not, how could we compile one?

I would take the dependency closure of the particular unikernel you're looking at building, e.g. the contents of `duniverse`.

**Comment by ehmry** [comment]
My knowledge is out of date then.
